specs-derive/src/impl_saveload.rs

```diff
@@ -0,0 +1,487 @@
+    let saveload_name = derive.saveload_name;
+    let saveload_generics = derive.saveload_generics;
+
+    let ser = derive.ser;
+    let de = derive.de;
+
+    let tt = quote!{
+        #[derive(Serialize, Deserialize, Clone)]
+        #[serde(bound = "MA: ::specs::saveload::Marker")]
+        pub #type_def
+
+        impl #impl_generics ::specs::saveload::IntoSerialize<MA> for #name #ty_generics #where_clause {
+            type Data = #saveload_name #saveload_generics;
+            type Error = ::specs::error::NoError;
+
+            fn into<F>(&self, mut ids: F) -> Result<Self::Data, Self::Error>
+            where
+                F: FnMut(::specs::Entity) -> Option<MA>
+            {
+                #ser
+            }
+        }
+
+        impl #impl_generics ::specs::saveload::FromDeserialize<MA> for #name #ty_generics #where_clause {
+            type Data = #saveload_name #saveload_generics;
+            type Error = ::specs::error::NoError;
+
+            fn from<F>(data: Self::Data, mut ids: F) -> Result<Self, Self::Error>
+            where
+                F: FnMut(MA) -> Option<Entity>
+            {
+                #de
+            }
+        }
+    };
+
+    // panic!("{}", tt);
+
+    tt
+}
+
+// Implements all elements of saveload common to structs of any type
+// (except unit structs like `struct Foo;` since they're auto-saveload);
+fn saveload_struct(
+    data: &mut DataStruct,
+    name: &mut Ident,
+    generics: &mut Generics,
+) -> SaveloadDerive {
+    use syn::Fields;
+
+    let mut saveload_generics = generics.clone();
+    saveload_generics.params.push(parse_quote!(MA));
+    match &mut saveload_generics.where_clause {
+        Some(clause) => clause.predicates.push(parse_quote!(
+            MA: ::serde::Serialize + ::serde::de::DeserializeOwned + ::specs::saveload::Marker,
+            for <'deser> MA: ::serde::Deserialize<'deser>
+        )),
+        clause => *clause = Some(parse_quote!(
+            where MA: ::serde::Serialize + ::serde::de::DeserializeOwned + ::specs::saveload::Marker,
+            for <'deser> MA: ::serde::Deserialize<'deser>
+        ))
+    }
+
+    let saveload_name = Ident::new(&format!("{}SaveloadData", name), name.span);
+
+    let saveload_fields: Vec<_> = data
+        .fields
+        .iter()
+        .cloned()
+        .map(|mut f| {
+            replace_entity_type(&mut f.ty);
+            f
+        })
+        .collect();
+
+    let (struct_def, ser, de) = if let Fields::Named(_) = data.fields {
+        saveload_named_struct(&name, &saveload_name, &saveload_generics, &saveload_fields)
+    } else if let Fields::Unnamed(_) = data.fields {
+        saveload_tuple_struct(
+            &data,
+            &name,
+            &saveload_name,
+            &saveload_generics,
+            &saveload_fields,
+        )
+    } else {
+        panic!("This derive does not support unit structs (Hint: they automatically derive FromDeserialize and IntoSerialize)");
```

Comment:
Ah now I get the above sentence; both places should describe it more like "every unit struct can derive `Serialize` and `Deserialize`, thus has a blanket impl fulfill their `IntoSerialize` / `IntoDeserialize`".

Reply:
Yeah, I think I wasn't thinking when I wrote this. TBH, we could probably just silently do nothing because of this instead of panicking. I think I decided to be explicit for some reason.